src/System.Private.ServiceModel/tests/Scenarios/Extensibility/WebSockets/WebSocketTests.cs

```diff
@@ -234,4 +234,230 @@ public static void WebSocketHttpsDuplexBinaryStreamed()
+            binding = new CustomBinding(textMessageEncodingBindingElement, httpsTransportBindingElement);
+
+            clientReceiver = new ClientReceiver();
+            context = new InstanceContext(clientReceiver);
+
+            channelFactory = new DuplexChannelFactory<IWSDuplexService>(context, binding, Endpoints.WebSocketHttpsDuplexTextStreamed_Address);
+            client = channelFactory.CreateChannel();
+
+            // *** EXECUTE *** \\
+            using (Stream stream = client.DownloadStream())
+            {
+                int readResult;
+                // Read from the stream, 1000 bytes at a time.
+                byte[] buffer = new byte[1000];
+                do
+                {
+                    readResult = stream.Read(buffer, 0, buffer.Length);
+                }
+                while (readResult != 0);
+            }
+
+            uploadStream = new FlowControlledStream();
+            uploadStream.ReadThrottle = TimeSpan.FromMilliseconds(500);
+            uploadStream.StreamDuration = TimeSpan.FromSeconds(1);
+
+            client.UploadStream(uploadStream);
+            client.StartPushingStream();
+            // Wait for the callback to get invoked before telling the service to stop streaming.
+            // This ensures we can read from the stream on the callback while the NCL layer at the service
+            // is still writing the bytes from the stream to the wire.  
+            // This will deadlock if the transfer mode is buffered because the callback will wait for the
+            // stream, and the NCL layer will continue to buffer the stream until it reaches the end.
+
+            Assert.True(clientReceiver.ReceiveStreamInvoked.WaitOne(ScenarioTestHelpers.TestTimeout), "Test case timeout was reached while waiting for the stream response from the Service.");
+            clientReceiver.ReceiveStreamInvoked.Reset();
+
+            // Upload the stream while we are downloading a different stream
+            uploadStream = new FlowControlledStream();
+            uploadStream.ReadThrottle = TimeSpan.FromMilliseconds(500);
+            uploadStream.StreamDuration = TimeSpan.FromSeconds(1);
+            client.UploadStream(uploadStream);
+
+            client.StopPushingStream();
+            // Waiting on ReceiveStreamCompleted from the ClientReceiver.
+            clientReceiver.ReceiveStreamCompleted.WaitOne();
+            clientReceiver.ReceiveStreamCompleted.Reset();
+
+            // *** VALIDATE *** \\
+            // Validation is based on no exceptions being thrown.
+
+            // *** CLEANUP *** \\
+            ((ICommunicationObject)client).Close();
+            channelFactory.Close();
+        }
+        finally
+        {
+            // *** ENSURE CLEANUP *** \\
+            ScenarioTestHelpers.CloseCommunicationObjects((ICommunicationObject)client, channelFactory);
+            clientReceiver.Dispose();
```

Comment:
Technically clientReceiver could still be null if we took an exception at the very beginning.
I wonder if we need something like ScenarioTestHelpers.DisposeAll() that takes a list of IDisposable objects (and it can test for null).